src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexRunnerFactory.cs

```diff
@@ -83,14 +83,46 @@ private sealed class Runner<TSetType> : RegexRunner where TSetType : notnull
 
             internal Runner(SymbolicRegexMatcher<TSetType> matcher) => _matcher = matcher;
 
-            protected override void InitTrackCount() { } // nop, no backtracking
+            protected internal override void Scan(Regex regex, ReadOnlySpan<char> text, int textstart, int prevlen, bool quick)
+            {
+                // Configure the additional value to "bump" the position along each time we loop around
+                // to call FindFirstChar again, as well as the stopping position for the loop.  We generally
+                // bump by 1 and stop at textend, but if we're examining right-to-left, we instead bump
+                // by -1 and stop at textbeg.
+                int stoppos = text.Length;
+                if (regex.RightToLeft)
+                {
+                    stoppos = 0;
+                }
 
-            protected override bool FindFirstChar() => true; // The logic is all in Go.
+                // If previous match was empty or failed, advance by one before matching.
+                if (prevlen == 0)
+                {
+                    if (textstart == stoppos)
+                    {
+                        runmatch = Match.Empty;
+                        return;
+                    }
 
-            protected override void Go()
-            {
-                ReadOnlySpan<char> inputSpan = runtext;
+                    runtextpos += regex.RightToLeft ? -1 : 1;
+                }
+
+                Go(text);
 
+                // If we got a match, we're done.
+                if (runmatch!._matchcount[0] > 0)
+                {
+                    if (quick)
+                    {
+                        runmatch = null;
```

Comment:
I'm still confused by nulling out runmatch.  quick means IsMatch is being used, which means a) the only thing that matters for an implementation is whether there is a match rather than where it is, which makes NonBacktracking's work easier, and b) that we're not handing out the Match object to anyone so it can be reused for the next invocation.  Both of those are quite important.

Reply:
This was done like this before because when quick == true, it expects Run() to return null when there is a Match, so overrides of Scan where doing this by setting runmatch to null. Now that we have moved all of the handling of runmatch into Run method, then this can be fixed.